src/server/worker/master.go

```diff
@@ -174,6 +199,14 @@ func (a *APIServer) jobSpawner(pachClient *client.APIClient) error {
 
 		switch {
 		case ppsutil.IsTerminal(jobInfo.State):
+			if jobInfo.StatsCommit != nil {
+				if _, err := pachClient.PfsAPIClient.FinishCommit(pachClient.Ctx(), &pfs.FinishCommitRequest{
```

Comment:
High-level question: should this be done by PFS? In other words, if a commit has no hashtree, would you expect downstream commits to have sensible output anyway? Maybe when you call `FinishCommit(commit, tree=nil)`, PFS should just close all downstream commits as well in a txn?

Reply:
Yeah, I actually had a similar thought while writing this. I think that would give us the semantics we currently have now, without the race conditions. @brycemcanally what do you think of this? It seems like this is the way to actually not have any of the race conditions we've been talking about.

Reply:
I think that makes sense. I am not too familiar with the transaction support, but if that makes it so that it only finishes both commits if they both have not been finished, then that should avoid the race condition where you call `StopJob` while the job is completing. I guess we would still need recovery code for the job state though.

Reply:
I'm opting to do this in a later PR, because I want to get this merged in for release tomorrow since it's proven to help a lot with merges getting stuck, and I think it's a big improvement over what we have now. Atomically finishing the downstream commits is a bit tricky to get passing tests. It's a simple implementation but there's a decent amount of testing that assumes things about the current behavior. I've opened #4266 to track this.